prover/src/constraints/composition_poly.rs

```diff
@@ -90,22 +101,27 @@ impl<E: FieldElement> CompositionPoly<E> {
 
 /// Splits polynomial coefficients into the specified number of columns. The coefficients are split
 /// in such a way that each resulting column has the same degree. For example, a polynomial
-/// a * x^3 + b * x^2 + c * x + d, can be rewritten as: (b * x^2 + d) + x * (a * x^2 + c), and then
-/// the two columns will be: (b * x^2 + d) and (a * x^2 + c).
-fn transpose<E: FieldElement>(coefficients: Vec<E>, num_columns: usize) -> Vec<Vec<E>> {
-    let column_len = coefficients.len() / num_columns;
-
-    let mut result = unsafe {
-        (0..num_columns)
-            .map(|_| uninit_vector(column_len))
-            .collect::<Vec<_>>()
+/// a * x^3 + b * x^2 + c * x + d, can be rewritten as: (c * x + d) + x^2 * (a * x + b), and then
+/// the two columns will be: (c * x + d) and (a * x + b).
+fn segment<E: FieldElement>(coefficients: Vec<E>, trace_len: usize) -> Vec<Vec<E>> {
+    let degree_plus_1 = degree_of(&coefficients) + 1;
+    let remainder = degree_plus_1 % trace_len;
+
+    let num_columns = if remainder == 0 {
+        degree_plus_1 / trace_len
+    } else {
+        (degree_plus_1 / trace_len) + 1
     };
 
+    let mut result = (0..num_columns)
+        .map(|_| vec![E::ZERO; trace_len])
+        .collect::<Vec<_>>();
+
     // TODO: implement multi-threaded version
-    for (i, coeff) in coefficients.into_iter().enumerate() {
-        let row_idx = i / num_columns;
-        let col_idx = i % num_columns;
-        result[col_idx][row_idx] = coeff;
+    for (i, coeff) in coefficients.iter().take(degree_plus_1).enumerate() {
+        let col_idx = i / trace_len;
+        let row_idx = i % trace_len;
+        result[col_idx][row_idx] = *coeff;
     }
```

Comment:
Ideally, we should be able to do this without re-allocating, but that may require unsafe code. For now, I wonder if we can use [split_off()](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.split_off) method and leave a TODO to come back and figure-out zero-allocation approach later.

Reply:
What about using slices like
```rust
let mut result = Vec::new();
    for i in (0..coefficients.len())
        .take(num_cols * trace_len)
        .step_by(trace_len)
    {
        let end = i + trace_len;
        let slice = &coefficients[i..end];
        result.push(slice.to_vec());
    }
```

Reply:
The above works! I wonder if we can simplify it like this:
```Rust
coefficients
    .chunks(trace_len)
    .take(num_cols)
    .map(|slice| slice.to_vec())
    .collect()
```

Reply:
Also, should we do something like:
```Rust
debug_assert!(degree_of(coefficients) < trace_len * nu_cols - 1);
```

Reply:
Makes sense!